lms/djangoapps/course_api/blocks/transformers/tests/test_block_completion.py

```diff
@@ -0,0 +1,119 @@
+"""
+Tests for BlockCompletionTransformer.
+"""
+from xblock.core import XBlock
+from xblock.completable import CompletableXBlockMixin, XBlockCompletionMode
+
+from lms.djangoapps.completion.models import BlockCompletion
+from lms.djangoapps.course_api.blocks.transformers.block_completion import BlockCompletionTransformer
+from lms.djangoapps.course_blocks.transformers.tests.helpers import ModuleStoreTestCase, TransformerRegistryTestMixin
+from student.tests.factories import UserFactory
+from xmodule.modulestore.tests.factories import CourseFactory, ItemFactory
+
+
+from ...api import get_course_blocks
+
+
+class StubAggregatorXBlock(XBlock):
+    """
+    XBlock to test behaviour of BlockCompletionTransformer
+    when transforming aggregator XBlock.
+    """
+    completion_mode = XBlockCompletionMode.AGGREGATOR
+
+
+class StubExcludedXBlock(XBlock):
+    """
+    XBlock to test behaviour of BlockCompletionTransformer
+    when transforming excluded XBlock.
+    """
+    completion_mode = XBlockCompletionMode.EXCLUDED
+
+
+class StubCompletableXBlock(XBlock, CompletableXBlockMixin):
+    """
+    XBlock to test behaviour of BlockCompletionTransformer
+    when transforming completable XBlock.
+    """
+    pass
+
+
+
+class BlockCompletionTransformerTestCase(TransformerRegistryTestMixin, ModuleStoreTestCase):
+    """
+    Tests behaviour of BlockCompletionTransformer
+    """
+    TRANSFORMER_CLASS_TO_TEST = BlockCompletionTransformer
+    COMPLETION_TEST_VALUE = 0.4
+
+    def setUp(self):
+        super(BlockCompletionTransformerTestCase, self).setUp()
+        self.user = UserFactory.create(password='test')
+
+
+    @XBlock.register_temp_plugin(StubAggregatorXBlock, identifier='aggregator')
+    def test_transform_gives_none_for_aggregator(self):
+        course = CourseFactory.create()
+        block = ItemFactory.create(category='aggregator', parent=course)
+        block_structure = get_course_blocks(
+            self.user, course.location, self.transformers
+        )
+
+        self._assert_block_has_proper_completion_value(
+            block_structure, block.location, None
+        )
+
+    @XBlock.register_temp_plugin(StubExcludedXBlock, identifier='excluded')
+    def test_transform_gives_none_for_excluded(self):
+        course = CourseFactory.create()
+        block = ItemFactory.create(category='excluded', parent=course)
+        block_structure = get_course_blocks(
+            self.user, course.location, self.transformers
+        )
+
+        self._assert_block_has_proper_completion_value(
+            block_structure, block.location, None
+        )
+
+    @XBlock.register_temp_plugin(StubCompletableXBlock, identifier='comp')
+    def test_transform_gives_value_for_completable(self):
+        course = CourseFactory.create()
+        block = ItemFactory.create(category='comp', parent=course)
+        BlockCompletion.objects.create(
```

Comment:
Try using [`BlockCompletion.objects.submit_completion(...)`](https://github.com/edx/edx-platform/blob/ab13544cf27132ff01f2ab1e615790e5bc29f436/lms/djangoapps/completion/models.py#L40) instead.

Reply:
Thanks! When I used `map_into_course` in the `transform()` method as you suggested in OC-3114, the problem is solved.

Reply:
@tomaszgy Great - but even still, I think we should still be using `submit_completion(...)` here, so that BlockCompletion objects get created consistently using that method everywhere in the codebase. You may also need to use [this decorator](https://github.com/edx/edx-platform/blob/c1e878fb778ba96567fd54d6a84f27ce3cdab5fc/lms/djangoapps/completion/test_utils.py#L12) on this test method, to enable the completion API.

Reply:
👍